pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java

```diff
@@ -121,8 +121,14 @@ public synchronized void addConsumer(Consumer consumer) throws BrokerServiceExce
 
     @Override
     public synchronized void removeConsumer(Consumer consumer) throws BrokerServiceException {
-        super.removeConsumer(consumer);
+        // The consumer must be removed from the selector before calling the superclass removeConsumer method.
+        // In the superclass removeConsumer method, the pending acks that the consumer has are added to
+        // messagesToRedeliver. If the consumer has not been removed from the selector at this point,
```

Comment:
> If the consumer has not been removed from the selector at this point

Looks like a race condition between sending messages to the consumer and remove the consumer from the selector?

We have a `synchronized` to protect the `readEntriesComplete` and `removeConsumer`. How can this happen?

Reply:
@codelipenghui Perhaps this happens if the messages to be redelivered is in the managed ledger cache. In this case,
```
readMoreEntries()
↓
readEntriesComplete()
↓
sendMessagesToConsumers()
```
are executed and completed synchronously in `PersistentDispatcherMultipleConsumers#removeConsumer()`.
https://github.com/apache/pulsar/blob/894d92b2be3bee334e7ce32760c4d2e7978603aa/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java#L184-L195

`synchronized` protection does not work because all of these methods are executed by the same thread.

Reply:
Ok, I think the stack trace here would be like: 

`removeConsumer()`
↓
`super.removeConsumer()`
↓
`readMoreEntries()` (This is what is used to trigger the re-delivery of the messages that were pending on the removed consumer)
↓
`readEntriesComplete()`
↓
`sendMessagesToConsumers()` As mentioned, this can select the removed consumer which is still in the selector list. 

At this point, the `sendMessagesToConsumers()` will fail and the message will stay into the `pendingAcks` set for that consumer, but, since the consumer was already removed, the redelivery of this message will not happen.

I think this change is the correct one.

Reply:
Oh, I see. Thanks for the explanation @massakam @merlimat. I missed that `super.removeConsumer()` also get a chance to call `readMoreEntries()`.